---
title: "DashboardReport"
author: "Jaclyn Vu"
date: "2023-05-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction 

#### Single Cell RNA Sequencing

For this project, a client approached our team with the desire to make navigating single cell RNA sequencing easier. Single-cell RNA sequencing (scRNA-seq) technology is used to unravel RNA transcripts within individual cells to attempt to identify the composition of varying cell types and function in different levels of organization in the body (i.e. tissues, organs, organisms) [(Jovic et al., 2022)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8964935/). Single cell sequencing shows the potential for applications in discovering new information, providing a better understanding of health and diseases, and developing a higher resolution catalog of cells in all living organisms, also known as an atlas [(Jovic et al., 2022)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8964935/).

#### Client Background 

The client investigated what the spatiotemporal transcriptional signatures underlying developmental trajectories in the Arabidopsis root were and what were the key transcriptional regulators in specifying and maintaining cell identity to regulate spatiotemporal organ development [(Shahan et al., 2022)](https://www.sciencedirect.com/science/article/pii/S1534580722000338?via%3Dihub). They also demonstrated the usefulness of an atlas by studying two mutant regulators, shortroot and scarecrow [(Shahan et al., 2022)](https://www.sciencedirect.com/science/article/pii/S1534580722000338?via%3Dihub). The big picture question that they were trying to answer is how transcriptional networks orchestrate organ development in multicellular organisms. They used the Arabidopsis root due to its simple structure and indeterminate growth, making it an ideal model [(Shahan et al., 2022)](https://www.sciencedirect.com/science/article/pii/S1534580722000338?via%3Dihub).

#### Client's Current Problem 

The client has predicted the cell clusters for their data, but they did not have a method of visualizing or numerically comparing the differences between the data. They were curious about which clusters are similar and different. They asked for some kind of application that allows them to quickly compare their cell clusters in a UMPA and PCA space. They also wanted a way to numerically compare which clusters were most similar and most different. In order to understand the biological meaning of the scRNA-seq, they wanted to visualize putative and previously identified biomarkers. Biomarkers are biological molecules found in body fluids or tissues that could be a sign of processes, conditions, or diseases [(National Cancer Institute, 2023)](https://www.cancer.gov/publications/dictionaries/cancer-terms/def/biomarker). Their team was able to predict new marker genes with patterns that differentiated from the clusters, so they wanted to know which clusters these new markers were expressed in. Lastly, they would like a visual representation of expression values based on their table of putative markers and previously known genes, and they would like to know if the predicted marker genes overlapped with the previously known biomarkers.

#### Client Provided Data 

The client provided our team with five CSV files. ‘AT_root_scRNAseq_UMAP.csv’ contains 3355 cell barcodes and their UMAP coordinates. ‘AT_root_scRNAseq_PCA.csv’ has 3355 cell barcodes and 50 principal component values. ‘AT_root_scRNAseq_clustermap.csv’ includes the 3355 cell barcodes and the corresponding cluster IDs, ranging from 0 to 20. There are 21 clusters. ‘AT_root_scRNAseq_putativemarkers.csv’ has the 21 gene IDs and the measured values in each of the 3355 cell barcodes. There is one gene per cluster. Lastly, ‘AT_root_scRNAseq_anticipatedmarkers.csv’ contains 40 gene IDs with the characterized expression patterns and the measured values in each of the 3355 cell barcodes.

#### Client Specifications

For the application the client requested five specifications. First, they wanted to display scRNAseq UMAP and PCA scatter plots in the same application tab, have two drop-down menus for selecting PC1 - PC50 vs PC1 - PC50 graph, and be able to update the PCA scatter plot without reloading. Second, there should be two drop-down menus for selecting PC1 - PC50 vs PC1 - PC50 and be able to update the PCA scatter plot without reloading. Third, a consistent color palette should be used in which each cell cluster has a unique color. Using a check-box menu, the user should be able to color UMAP or PCA with cell cluster identity. There should be a check-box for each cluster, and we needed to provide an additional check-box to hide cells that were not currently selected. Fourth, on a separate application tab, a bar chart that counts the number of cells in each cell cluster and a heat map of Pearson correlation in an all vs all comparison of cell clusters based on cluster average PCA values should be included. Lastly, on a third application tab, the UMAP data with a drop-down menu to select from the predicted or previously known markers should be displayed, and when a marker is selected, UMAP points should be re-colored according to the scaled gene expression value in all barcodes.

## Methods

#### Initial Approach 

To meet our client’s needs for single cell cluster visualization and quantitative comparison, our team made a goal to have the functional components of our application by May 19th. In order to complete our build of the app by this deadline, each team member focused on one component of the application, and aimed to have rough drafts of each component done by Tuesday, May 16th. From there, we reviewed each others’ components and incorporated feedback from each other into the final dashboard, which we compiled by May 19th. As a team, we developed a plan to achieve each task, and then, we delegated the tasks to each team member. The following were the ideas we initially developed to achieve all of the client’s specifications.

The first task was to compare the cell type clusters to see which are most similar and different. Our team wanted to make a scatter plot to display the scRNAseq UMAP and PCA dimensions. These plots should use the files “AT_root_scRNAseq_UMAP.csv” and “AT_root_scRNAseq_PCA.csv” for the data. For the UMAP visualization, the input for the plotting functions was the x and y columns. The user could choose which principal components they want to plot against each other. In addition to the PCA scatter plot, the plot should be annotated with how much variance is explained by the first two principal components. These annotations would give the user an idea of how effective the plot is at representing the full-high dimensional data set. If we assumed that the first 50 principal components would be able to explain most of the variance, this could be calculated. In addition to the plots, accompanying texts and commentary should be used to interpret the percentage of variance. Lastly, the commentary on the UMAP, regarding its strengths and weaknesses in the representation for the full data, should be provided in the accompanying text. This would ideally help our client compare the cell type clusters.

The second task was to make the dropdown menus for selecting PC1 - PC50 vs. PC1 - PC50. We planned to use the “selectInput()” function. Based on the user’s selection, the app should update the PCA without reloading. This helps aid in comparing the cell type clusters, but it specifically allows the client to visualize the comparison of the PCA transcriptome in detail.

For the third task, we explored using coolors.co to generate a color palette ourselves by collecting the hex color codes to keep track of the colors we would use for each cell cluster and the cell cluster identities within the UMAP and PCA plots. With a check-box for each cluster and a check-box menu, users should be able to select which colors are turned on and off. There should also be a checkbox to hide the cells that are not in the currently selected clusters. Another possibility we explored was the use of themes within Shiny. These approaches would bring consistency and cleanliness for the user interface experience, making interpreting easier as well. These plans should allow for a consistent color palette for the user interface.

For the fourth task, in the second application tab, there should be a bar chart that counts the number of cells in each cell cluster and a heatmap of the Pearson correlation in an all vs. all comparison of cell clusters based on cluster average PCA values. The bar chart should be built with the “barplot()” function to compare the cell clusters and the function “pheatmap()” could be used to generate a heatmap. The heatmap should be able to better illustrate the volume of points within each cluster, and the bar chart would display the relationship between the clusters. Using the “mean()” function, the cluster average of the PCA values would be calculated, and using “cor.test()”, to compare the correlation values between plots, the Pearson correlation values would be calculated. With the barchart, heatmap, and the calculated values, this should help the clients understand cell differentiation.

For the fifth task, in a third application tab, the user should be able to use a dropdown menu to select from the predicted or previously known markers from the UMAP data. This could be completed by using the “selectInput()” function to make the dropdown menu. We would load the UMAP package with “library(UMAP)”, and we would apply the transformation with “df.umap <- umap(df.data)” in order to display the UMAP data, By setting the specific markers to a boolean values, we would use the hovering mouse over a marker or clicks on a specific marker to we will mark a boolean value as true. This would initiate the color change of the UMAP points according to scaled gene expression value in all barcodes. By doing this, the client should be able to see which clusters putative marker genes are in, visualize which genes meet the statistical threshold for differentiating clusters, and see what overlap is present between the putative markers and their list of previously known biomarkers.

#### My Contribution 

In order to compare UMAP and PCA transcriptome visualization techniques, our team needed tools to plot both types of data. I was the head for the first task of displaying the scRNAseq and PCA scatter plots in the same application tab. To begin, I loaded the shiny package with “library(shiny).” I wanted to map out the layout of the application so that it would be easier to assemble the team member’s code all together into one chunk of code. To build the fluid page, I included a title panel, made three tabs and titled them according to the information that would later be displayed. For the first tab, I included two rows and two columns so that I could put a title for each figure and a caption about each graph. I used the function “plotOutput’’ to create the variables that would later be defined to create the graphs. The names I chose were “umap” and “pca.” This step provided the structure of our Shiny application.

Within the “server()” function, I knew my tasks were to graph figures, so I first loaded the ggplot2 package with “library(ggplot2).” I then went ahead and imported the required CSV files for my tasks, “AT_root_scRNAseq_UMAP.csv” and “AT_root_scRNAseq_PCA.csv.” For the UMAP and PCA data, I used the function, ”read.csv(),” and I saved each into a variable. The UMAP data was visualized with a simple scatter plot since we only wanted to compare UMAP_1 and UMAP_2. In order to render the plot for “umap,” I accessed the “umap” variable created in the fluid page with “output$umap” and setted it equal to the function “renderPlot({}).” Within this plot, I used the function, “ggplot()” to create a simple scatter plot. The parameters I included were the origin of the data, the aesthetic mapping (“aes(x,y,…)”), and the “geom_point()” function to create the scatter plot, producing the following figure:

```{r}
umap_df <- read.csv("./AT_root_scRNAseq_UMAP.csv")
library(ggplot2)
ggplot(data = umap_df, aes(UMAP_1, UMAP_2)) + geom_point() + labs(x="UMAP Dimension 1", y="UMAP Dimension 2",
             title="UMAP Projection of Single Cell Expression Profiles") # simple scatterplot
```

From the output, I was able to successfully plot the UMAP data in its entirety. We still needed to include a checkbox menu for the user to select the clusters they would like to see and or hide and to add was the color coordination from the cluster, which Dan, a team member, implemented these features later on. In combination, we were able to visualize the UMAP transcriptome.

For the PCA plot, I realized that this graph would be more complicated as the user should be able to select PC1 through PC50 for each axis, which is task two as described earlier. I asked Connor, the team member responsible for that assignment, if they would like me to work on it, and he could potentially help with another task. This ended up working out since a team member needed help with their portion of the project. I created a basic scatter plot, simply comparing PC1 vs PC2 to get a plot on the page, shown below.

```{r}
pca_df <- read.csv("./AT_root_scRNAseq_PCA.csv")
library(ggplot2)
pca_data <- data.frame (PC_x  = pca_df['PC_1'],
                        PC_y = pca_df['PC_2']
                  )
ggplot(data = pca_data, aes(PC_1, PC_2)) + geom_point() + labs(title = "PC_2 vs. PC_1",  x = "PC_1", y = "PC_2") # simple scatterplot
```

This figure is not the final figure that I wanted implemented into the data dashboard, but it gave me a starting point to visualizing the PCA transcriptome. Rather than hard coding in the variables to compare, I needed to make them variables and make the graph reactive to the user’s choices in order to visualize the PCA transcriptome. In order to do this, I used the same approach for the UMAP plot. However, before plotting I needed to create a data frame with “data.frame()” to pull the desired columns of the PCA values to plot. I called the “ggplot()” function the same way except I used “aes_string()” instead of “aes()” since I was passing variables that contained the x and y variable names. In order to allow the user to input the desired PCA values, I created a numerical input box for both x and y values. The range was between 1 and 50 with a default value of 1. The variable names were “pca_x” and “pca_y” respectively. Since the plot should update without reloading, I knew I had to make a reactive graph dependent on the user’s selections. I created and assigned the function “pca_plot” to a “reactive({})” call. We accessed the user’s inputs with “input$pca_x” and “input$pca_y.” The function “paste0()” was used to combine “PC_” and the number the user selected into one string. This string was saved into variables “x_val” and “y_val.” We created the variable “pca_data” to save a data frame with the user selected columns. The final call in the reactive function was the “ggplot()” function with the “pca_data,” “x_val” and “y_val” in the “aes_string()” function, the “geom_point()” function call, and we used “labs()” to add a title to the graph with what the plot is comparing, using the “paste0()” function and the “x_val” and “y_val” variables. Finally, the “output$pca” was set equal to the “renderPlot()” function, containing the “pca_plot()” function we created. Thus, task one and two were completed. Dan, the member assigned with keeping the color pallet consistently, implemented their code into the r code I produced.

As mentioned earlier, after following up with the team members, we had a sick team member, Morgan, that needed help completing their task. Since I had completed my task along with Connor’s original task, together, we decided to work on task three together. In order to meet the deadline that we set, I picked up an extra task. I worked on the bar chart that counts the number of cells in each cell cluster. With the aid of our boss, Sam, I was able to complete this bar chart.

```{r}
clustermap <- read.csv("./AT_root_scRNAseq_clustermap.csv")
library(ggplot2)
library(dplyr)
grouping_df <- clustermap %>% group_by(seurat_clusters) %>% 
      summarise(total_count=n(), .groups = 'drop')
# Barplot
ggplot(grouping_df, aes(seurat_clusters, total_count)) + 
      geom_bar(stat = "identity") + labs(title = "Number of Cells vs. Seurate Cell Cluster", x = "Seurate Cluster ID", y = "Number of Cells")
```

This bar chart successfully compared the number of cells to the seurate clusters without the color coordination, which again, Dan would later implement. The CSV with the clustermap data was read in and the “dplyr” package was loaded. We created a grouped data frame, “grouping_df” with the total number of observations in each seurate cluster from the cluster map data. This data was forwarded in a chain with the forward pipe operator (“%>%”). The grouping information was then dropped so that the resulting data frame would not include the clusters as individual columns. The function “ggplot” was used to graph the barchart. The parameters included the “grouping_df”, “aes()” function call with the axis information, the function “geom_bar()” to create the bar chart with the parameter “stat = identity” so that the height of the bars will be determined by the data provided, and lastly a “labs()” call for the plot and axis titles for the graph. Again, Dan implemented color coordination into the graphs afterwards.

## Conclusion 

The client is having trouble navigating the data from single cell RNA seq. They were able to predict the cell clusters in their data, but they want a way to visually or numerically compare the differences between these clusters. They want to see which clusters are similar and which are different. To see how similar or different the cell clusters are in their data, we created a visualization using colors to represent each cluster. In order to quickly compare the UMAP and PCA space, we built two plots side-by-side in one application tab and used the same color coordination to allow for quick and easy comparisons. In the future, these graphs may benefit by being plotted into a higher dimensional space. When looking at the UMAP graph alone. The distance between the points are slim and are difficult to differentiate. If we added the function to zoom and rotate the graph, there could potentially improve the user experience by adding more exploration and interactivity in the graphs. The input method of choosing the axes does not look the cleanest. Some other approaches could be a sliding bar for each axis or a drop-down menu. Similarly with the PCA visualization, there are often a lot of points on the graph. It is difficult to speak in regard to all combinations, but since the user is comparing the two visualizations, it would be nice to add the feature to both graphs. The PCA figure is nice to compare the different PCA dimensions if the user has specific PC values they want to compare, but there may need to be some exploration on different ways to visualize the different combinations of PCA transcriptomes.

In addition, our client wanted to know how many cells are in each cell cluster in order to understand cell differentiation. The bar chart that was made to compare the number of cells to the clusters was fairly simple and easy to understand, but the data suggests that only the somewhat common cell types are represented. It would be interesting to include more rare types of cells in the data set. Dan was able to add the color coordination to the bar chart as well, which made the bar chart more fun to look at, but otherwise, this was a fairly straightforward figure.
